Feature/Optimize SAI.getPostQueryOrdering as SAI.postQuerySort

marianne-manaog opened the pull request:
This PR aims to address [this comment](https://github.com/datastax/cassandra/pull/690#issuecomment-1681255839) from the previous [PR no. 690](https://github.com/datastax/cassandra/pull/690). 

The post-query sort is doing a comparison-based sort where we compute the similarity to the query vector for each comparison, which is expensive. This PR aims to convert it to a decorate-sort-undecorate pattern.

michaeljmarshall commented on src/java/org/apache/cassandra/cql3/ResultSet.java:
```diff
@@ -47,7 +47,7 @@ public class ResultSet
 
     public final ResultMetadata metadata;
-    public final List<List<ByteBuffer>> rows;
+    public List<List<ByteBuffer>> rows;
```
Nit: I think we can keep this as a `final` variable.

marianne-manaog replied:
Thanks - I addressed it [here](https://github.com/datastax/cassandra/pull/710/commits/fff4a3327f31f2a812bd8bc77472ba423005c860).

michaeljmarshall commented on test/unit/org/apache/cassandra/index/sai/v1/StorageAttachedIndexTest.java:
```diff
@@ -0,0 +1,166 @@
+    @Test
+    public void testOrderResults() {
+        ResultSet.ResultMetadata resultMetadata = new ResultSet.ResultMetadata(new ArrayList<>(cfs.metadata.get().columns()));
+        ResultSet resultSet = new ResultSet(resultMetadata);
+        QueryOptions queryOptions = QueryOptions.create(ConsistencyLevel.ONE,
+                                                        byteBufferList,
+                                                        false,
+                                                        PageSize.inRows(1),
+                                                        null,
+                                                        null,
+                                                        ProtocolVersion.CURRENT,
+                                                        KEYSPACE);
+
+        List<List<ByteBuffer>> rows = new ArrayList<>();
+        rows.add(byteBufferList);
+        resultSet.rows = rows;
```
Nit: I was able to get the test to pass with this modification, which also means we can make the `row` field `final`.

```suggestion
        QueryOptions queryOptions = QueryOptions.create(ConsistencyLevel.ONE,
                                                        byteBufferList,
                                                        false,
                                                        PageSize.inRows(1),
                                                        null,
                                                        null,
                                                        ProtocolVersion.CURRENT,
                                                        KEYSPACE);
        List<List<ByteBuffer>> rows = new ArrayList<>();
        rows.add(byteBufferList);
        ResultSet resultSet = new ResultSet(resultMetadata, rows);
```

marianne-manaog replied:
Thanks for this - I agree and addressed it [here](https://github.com/datastax/cassandra/pull/710/commits/fff4a3327f31f2a812bd8bc77472ba423005c860).

michaeljmarshall commented on src/java/org/apache/cassandra/cql3/statements/SelectStatement.java:
```diff
@@ -115,6 +114,9 @@ public class SelectStatement implements CQLStatement.SingleKeyspaceCqlStatement
 
     private final boolean isReversed;
 
+    private static int columnIndex;
+    private static SingleRestriction restriction;
```
These are declared, but never set. Where should we set them? Also, we likely don't want them to be `static` fields since they will be different for each `SelectStatement` instance.

michaeljmarshall replied:
Most of the vector related tests are failing because the `restriction` is null.

```
java.lang.NullPointerException
	at org.apache.cassandra.cql3.statements.SelectStatement.orderResults(SelectStatement.java:1027)
	at org.apache.cassandra.cql3.statements.SelectStatement.process(SelectStatement.java:904)
	at org.apache.cassandra.cql3.statements.SelectStatement.processResults(SelectStatement.java:520)
	at org.apache.cassandra.cql3.statements.SelectStatement.executeInternal(SelectStatement.java:544)
	at org.apache.cassandra.cql3.statements.SelectStatement.executeLocally(SelectStatement.java:526)
	at org.apache.cassandra.cql3.statements.SelectStatement.executeLocally(SelectStatement.java:101)
	at org.apache.cassandra.cql3.QueryProcessor.executeInternal(QueryProcessor.java:483)
	at org.apache.cassandra.cql3.CQLTester.executeFormattedQuery(CQLTester.java:1368)
	at org.apache.cassandra.cql3.CQLTester.execute(CQLTester.java:1356)
	at org.apache.cassandra.index.sai.cql.VectorAndLuceneTest.partitionDeleteVectorInMemoryTest(VectorAndLuceneTest.java:86)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater$1.execute(IdeaTestRunner.java:38)
	at com.intellij.rt.execution.junit.TestsRepeater.repeat(TestsRepeater.java:11)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:35)
	at com.intellij.rt.junit.JUnitForkedStarter.main(JUnitForkedStarter.java:29)
```

marianne-manaog replied:
I removed them and the tests seem to be passing now, as per the refactored `orderResults` method at [this commit](https://github.com/datastax/cassandra/pull/710/commits/22cf12dc806b99372f602ec7f7aff702fb727191).

michaeljmarshall reviewed (commented):
Leaving some minor feedback from my first pass on the PR.

I need to look closer at the referenced PRs to understand the context for this PR. I'll be able to give a more thorough review later tonight or tomorrow morning at the latest.

michaeljmarshall commented on src/java/org/apache/cassandra/cql3/statements/SelectStatement.java:
```diff
@@ -1017,14 +1019,14 @@ private boolean needIndexOrdering()
 
-        Comparator<List<ByteBuffer>> comparator = orderingComparator.prepareFor(table, options);
-        if (comparator != null)
-            Collections.sort(cqlRows.rows, comparator);
```
It looks like the sort is removed here but not added somewhere else.

jbellis replied:
I assumed we already have a test that exercises this, is that not the case?

michaeljmarshall replied:
I assume we have a test for that too. I checked out the branch, and any relevant vector tests are failing right now due to an NPE when calling `restriction.findSupportingIndex`, which is before exercising the `postQuerySort` method. Is the goal to move sorting into the `postQuerySort` method?

marianne-manaog replied:
Hi @michaeljmarshall, thanks for checking that. I moved the sorting into the `postQuerySort` method via the DSU pattern as per [this commit](https://github.com/datastax/cassandra/pull/710/commits/22cf12dc806b99372f602ec7f7aff702fb727191).

michaeljmarshall reviewed (commented):
Which part of the `getPostQueryOrdering` method are we trying to optimize? A little more description of the problem and the solution could be helpful. Thanks @marianne-manaog!

michaeljmarshall commented:
Will that solution collide with the work in https://github.com/riptano/VECTOR-SEARCH/issues/9? If we attach the computed score to the row, like is done [here](https://github.com/datastax/cassandra/compare/vsearch...datastax:cassandra:VECTOR-20?expand=1), we won't need to decorate because we already have the score. I am still getting my bearings, so perhaps I am missing something.

jbellis commented:
Yes, if we had VS-9 today we wouldn't need this.  But this is a much smaller change, and it won't make VS-9 harder, so it's fine to start here.

michaeljmarshall commented:
Great, that definitely makes sense.

marianne-manaog commented on src/java/org/apache/cassandra/index/sai/StorageAttachedIndex.java:
```diff
@@ -561,7 +561,10 @@ public Comparator<List<ByteBuffer>> getPostQueryOrdering(Restriction restriction
+
+            Pair<float[], Double> decoratedLeft = Pair.create(left, scoreLeft);
+            Pair<float[], Double> decoratedRight = Pair.create(right, scoreRight);
+            return Double.compare(decoratedRight.right, decoratedLeft.right); // descending order
```
Hi @michaeljmarshall, that is right. I moved the decorate part out of the previous comparator block as per the refactored `postQuerySort` method via the DSU pattern as per [this commit](https://github.com/datastax/cassandra/pull/710/commits/22cf12dc806b99372f602ec7f7aff702fb727191).

michaeljmarshall commented on src/java/org/apache/cassandra/index/sai/StorageAttachedIndex.java:
```diff
@@ -542,26 +543,36 @@ public RowFilter getPostIndexQueryFilter(RowFilter filter)
-    public Comparator<List<ByteBuffer>> getPostQueryOrdering(Restriction restriction, int columnIndex, QueryOptions options)
-    {
-        // For now, only support ANN
+    public void postQuerySort(ResultSet cqlRows, Restriction restriction, int columnIndex, QueryOptions options) {
```
Nit: I think the code formating has the `{` on the new line. Also, we can probably keep the comment about only supporting ANN for now.

marianne-manaog replied:
Thanks. Addressed at [7e0c233](https://github.com/datastax/cassandra/pull/710/commits/7e0c2338a160fafc3b22ec5d90fac79cfdffb0ce).

michaeljmarshall commented on src/java/org/apache/cassandra/index/sai/StorageAttachedIndex.java:
```diff
@@ -542,26 +543,36 @@ public RowFilter getPostIndexQueryFilter(RowFilter filter)
+                                                                            ByteBuffer vectorBuffer = row.get(columnIndex);
+                                                                            float[] vector = TypeUtil.decomposeVector(indexContext, vectorBuffer.duplicate());
+                                                                            Double score = (double) similarityFunction.compare(vector, targetVector);
+                                                                            return new Pair<>(vectorBuffer.duplicate(), score);
```
I think we want to store the `row` instead of the `vectorBuffer`. That way, when we undecorate the sorted list, we just have to call `pair.left`.

marianne-manaog replied:
Great catch, thanks. Addressed at [f10da6c](https://github.com/datastax/cassandra/pull/710/commits/f10da6c38a9271d5aff49da93b24b5e6649e8d08).

michaeljmarshall reviewed (commented):
Great progress @marianne-manaog. This is very close to completion. I left a couple comments on things I think we should consider.

marianne-manaog commented on src/java/org/apache/cassandra/utils/Pair.java:
```diff
@@ -24,7 +24,7 @@ public class Pair<T1, T2>
     public final T2 right;
 
-    protected Pair(T1 left, T2 right)
+    public Pair(T1 left, T2 right)
```
Thanks. I addressed this at [a0a1cbc](https://github.com/datastax/cassandra/pull/710/commits/a0a1cbc07b1beba6e16341fc0926aad90531f844).

jbellis replied:
This is because Pair predates Cassandra's adoption of 1.7 and the diamond operator.

marianne-manaog commented:
> Great progress @marianne-manaog. This is very close to completion. I left a couple comments on things I think we should consider.

Thank you @michaeljmarshall for this further review. I addressed your comments and please let me know if you have any further questions or concerns.

michaeljmarshall commented on test/unit/org/apache/cassandra/index/sai/v1/StorageAttachedIndexTest.java:
Nit: this file could be moved to `test/unit/org/apache/cassandra/index/sai/StorageAttachedIndexTest.java` (up a directory)

marianne-manaog replied:
Addressed at [ed37515](https://github.com/datastax/cassandra/pull/710/commits/ed37515c171201090558d74524ac91a7b5b30e4d)

michaeljmarshall reviewed (approved):
LGTM

marianne-manaog commented on src/java/org/apache/cassandra/index/sai/StorageAttachedIndex.java:
```diff
@@ -542,26 +543,35 @@ public RowFilter getPostIndexQueryFilter(RowFilter filter)
+        // Decorate-sort-undecorate to optimize sorting of vectors by their similarity scores
+        List<Pair<List<ByteBuffer>, Double>> listPairsVectorsScores = buffRows.stream()
+                                                                              .map(row -> {
+                                                                                  List<ByteBuffer> vectorRow = new ArrayList<>(row);
```
I wanted to convert each `row`/ByteBuffer in `buffRows` into a `Pair` that includes the initial List<ByteBuffer> and the computed similarity score.

michaeljmarshall replied:
Does it work to just pass the `row` in as the left argument to the Pair?

marianne-manaog replied:
Thanks for your clarification and suggestion. Addressed at [a3ee879](https://github.com/datastax/cassandra/pull/710/commits/a3ee87991d4b1d6470efb75bdc7faafc1df89ba8)
